topbin/coqnative_bin.ml

```diff
@@ -0,0 +1,318 @@
+  Nativelib.compile_library ast fn
+
+let get_used_load_paths () =
+  String.Set.elements
+    (DPmap.fold (fun m f acc -> String.Set.add
+      (Filename.dirname f) acc)
+       !libraries_table String.Set.empty)
+
+let _ = Nativelib.get_load_paths := get_used_load_paths
+end
+
+let add_path ~unix_path:dir ~coq_root:coq_dirpath =
+  let open System in
+  if exists_dir dir then
+    begin
+      Loadpath.add_load_path (dir,coq_dirpath)
+    end
+  else
+    Feedback.msg_warning (str "Cannot open " ++ str dir)
+
+let convert_string d =
+  try Id.of_string d
+  with CErrors.UserError _ ->
+    Flags.if_verbose Feedback.msg_warning
+      (str "Directory " ++ str d ++ str " cannot be used as a Coq identifier (skipped)");
+    raise Exit
+
+let coq_root = Id.of_string "Coq"
+
+let add_rec_path ~unix_path ~coq_root =
+  let open System in
+  if exists_dir unix_path then
+    let dirs = all_subdirs ~unix_path in
+    let prefix = DirPath.repr coq_root in
+    let convert_dirs (lp, cp) =
+      try
+        let path = List.rev_map convert_string cp @ prefix in
+        Some (lp, Names.DirPath.make path)
+      with Exit -> None
+    in
+    let dirs = List.map_filter convert_dirs dirs in
+    List.iter Loadpath.add_load_path dirs;
+    Loadpath.add_load_path (unix_path, coq_root)
+  else
+    Feedback.msg_warning (str "Cannot open " ++ str unix_path)
+
+let init_load_path includes =
+  let () = Envars.set_coqlib ~fail:(fun x -> CErrors.user_err Pp.(str x)) in
+  let ( / ) = Filename.concat in
+  let coqlib = Envars.coqlib () in
+  let user_contrib = coqlib/"user-contrib" in
+  let xdg_dirs = Envars.xdg_dirs in
+  let coqpath = Envars.coqpath in
+  let plugins = coqlib/"plugins" in
+  (* NOTE: These directories are searched from last to first *)
+  (* first standard library *)
+  add_rec_path ~unix_path:(coqlib/"theories") ~coq_root:(Names.DirPath.make[coq_root]);
+  (* then plugins *)
+  add_rec_path ~unix_path:plugins ~coq_root:(Names.DirPath.make [coq_root]);
+  (* then user-contrib *)
+  if Sys.file_exists user_contrib then
+    add_rec_path ~unix_path:user_contrib ~coq_root:Loadpath.default_root_prefix;
+  (* then directories in XDG_DATA_DIRS and XDG_DATA_HOME *)
+  List.iter (fun s -> add_rec_path ~unix_path:s ~coq_root:Loadpath.default_root_prefix)
+    (xdg_dirs ~warn:(fun x -> Feedback.msg_warning (str x)));
+  (* then directories in COQPATH *)
+  List.iter (fun s -> add_rec_path ~unix_path:s ~coq_root:Loadpath.default_root_prefix) coqpath;
+  (* then current directory *)
+  add_path ~unix_path:"." ~coq_root:Loadpath.default_root_prefix;
+  (* additional loadpath, given with -R/-Q options *)
+  List.iter
+    (fun (unix_path, coq_root) -> add_rec_path ~unix_path ~coq_root)
+    (List.rev includes)
+
+let fb_handler = function
+  | Feedback.{ contents; _ } ->
+    match contents with
+    | Feedback.Message(_lvl,_loc,msg)->
+      Format.printf "%s@\n%!" Pp.(string_of_ppcmds msg)
+    | _ -> ()
+
+let init_coq () =
```

Comment:
Is there a good reason not to use sysinit?

Reply:
Same reason the checker doesn't use it: it's too high in the dependency chain.

Reply:
The checker has the, to me good, reason to not depend on the higher layers.
I don't see how this very same reason applies here, but maybe there are others.

Reply:
More precisely, these is some code about parsing and interpreting -R, -Q, -I here which has no reason to be duplicate.

Reply:
It has the same semantics as the checker, i.e. `-R / -Q` are identified since relative paths are not used. I believe @ejgallego 's work is precisely going to extract the relevant parts from coqinit to put them at the lowest point possible so that all Coq binaries can use them as a library. Otherwise your complaint also applies to coqchk and coqdep.

Also, the vast majority of flags accepted in coqinit are totally irrelevant for that tool.

Reply:
Fair enough, but then maybe we should this PR depend on that one. (It is a new command and is still a draft)
@ejgallego what is the status of the split of sysinit ?

Reply:
See https://github.com/coq/coq/pull/13287#issuecomment-808747055

Reply:
So you guys think it is a good idea to make a dune release which calls a tool merged in coq master ten minutes before?

I'm find merging as is and clean up later, but doing it for that reason is ********

Reply:
I personally don't care about the dune release, I just want the coqnative binary to be shipped in the next Coq release. Even without support from any build system it would help some users who repeatedly asked for features that would be solved with coqnative.

Reply:
@gares actually there are two separate things here:
- unifying coqchk / coqdep / coqnative code: I have a PR doing that using a sites/findlib inspired API which IMHO is quite Cool as avoid misuses by making paths opaque, but that depends on #13617 [for bootstrap reasons, this refactoring work is super hard to with out make-based OCaml build system]
- having dune support coqnative, this is largely orthogonal and not a problem, as dune can check if coqnative is present in the workspace and enable the rules conditionally on that.

Reply:
My point of view as whether to merge is that any delay to a PR usually has an exponential cost, it is way more easier for everyone to merge as soon a possible.

Reply:
For this PR I am very happy to have it soon in master and we can tweak it, than having a complex layer of inter-depending PRs

Reply:
> @ejgallego what is the status of the split of sysinit 

@gares I think I am mostly done with the "use dune to build OCaml" so will submit a draft PR pretty soon; I'd personally would merge whichever is ready first, this tool will indeed use the same code than coqchk.